### ISSUE TITLE
[Question] Is it possible to mock specific parameters of a cmdlet?

### ISSUE THREAD

**Issue by markwragg** [pester/Pester#771]
I was looking at StackOverflow Pester questions and came across [this one](http://stackoverflow.com/q/40221060/2796058) from a few months back which was looking to Mock `Out-File` such that they could redirect the output to a test location. [My answer](http://stackoverflow.com/a/43585690/2796058) suggests using `Set-Content` in a Mock as that has equivalent functionality. However this probably isn't an ideal solution, but seemed the only way because calling a cmdlet from within a Mock that is Mocking it just seemed to call the Mock and create an ugly infinite loop.

Is there a right/possible way to utilize the legitamate cmdlet within code that is Mocking it, or some other way to override specific parameters of a cmdlet/function rather than it entirely? Or is this just out of scope behavior for Pester?

**Comment by dlwyatt** [comment]
You can always call the original command from within a mock.  For example, you could save the original command to a variable before mocking it, and execute that with the call operator (`&`):

```posh
Describe Whatever {
    $outFile = Get-Command Out-File
    Mock Out-File { & $outFile -InputObject $InputObject -FilePath Some\Mocked\Path.txt -Append:$Append -Force:$Force -NoClobber:$NoClobber -Width:$Width -NoNewline:$NoNewline }
}
```

It's a bit wordy to pass on all the original parameters, but unfortunately, you don't have an easy $PSBoundParameters table in a mock to work with, due to implementation details.  (If you don't mind risking breaking changes later, it is possible to make that work, but we don't support it or guarantee that it'll continue to work later:)

```posh
Describe Whatever {
    $outFile = Get-Command Out-File
    Mock Out-File {
        $parentParams = Get-Variable PSBoundParameters -Scope 1 -ValueOnly
        $boundParams = $parentParams['___BoundParameters___'].Clone()

        $boundParams['FilePath'] = 'Some\Mocked\Path.txt'
        & $outFile @boundParams
    }
```

**Comment by markwragg** [comment]
This makes a lot of sense and seems like a good solution.. Thanks for the quick response!
